src/QsCompiler/SyntaxProcessor/TypeInference/InferenceContext.fs

```diff
@@ -20,100 +20,143 @@ open Microsoft.Quantum.QsCompiler.Transformations.QsCodeOutput
 open Microsoft.Quantum.QsCompiler.Utils
 
 /// A placeholder type variable.
-type private Variable =
+type Variable =
     {
         /// The substituted type.
         Substitution: ResolvedType option
-
         /// The list of constraints on the type of this variable.
         Constraints: Constraint list
-
         /// Whether this variable encountered a type inference error.
         HasError: bool
-
         /// The source range that this variable originated from.
         Source: Range
     }
 
-module private Variable =
+module Variable =
     /// Adds a type constraint to the type of the variable.
     let constrain typeConstraint variable =
         { variable with Constraints = typeConstraint :: variable.Constraints }
 
-/// An ordering comparison of two types.
-type private Ordering =
-    /// The type is a subtype of the other type.
+/// An ordering comparison between types.
+type Ordering =
+    /// The type is a subtype of another type.
     | Subtype
-
     /// The types are equal.
     | Equal
-
-    /// The type is a supertype of the other type.
+    /// The type is a supertype of another type.
     | Supertype
 
-module private Ordering =
-    /// Negates the direction of the ordering.
-    let not =
+module Ordering =
+    /// Reverses the direction of the ordering.
+    let reverse =
         function
         | Subtype -> Supertype
         | Equal -> Equal
         | Supertype -> Subtype
 
-/// Remembers context when recursively comparing two types side-by-side.
-type private TypeContext =
-    {
-        /// The left-hand type.
-        Left: ResolvedType
+type 'a Relation = Relation of lhs: 'a * ordering: Ordering * rhs: 'a
+
+module Relation =
+    let map f (Relation (x, ordering, y)) = Relation(f x, ordering, f y)
 
-        /// The right-hand type.
-        Right: ResolvedType
+module RelationOps =
+    let (<.) lhs rhs = Relation(lhs, Subtype, rhs)
 
-        /// The most recent relevant ancestor of the left-hand type.
-        OriginalLeft: ResolvedType
+    let (.=.) lhs rhs = Relation(lhs, Equal, rhs)
 
-        /// The most recent relevant ancestor of the right-hand type.
-        OriginalRight: ResolvedType
+    let (.>) lhs rhs = Relation(lhs, Supertype, rhs)
+
+type TypeMismatch =
+    {
+        Expected: ResolvedType
+        ExpectedContext: ResolvedType option
+        Actual: ResolvedType
+        ActualContext: ResolvedType option
```

Comment:
Some comments here explaining the fields, especially the context fields, would be helpful. I'm not yet understanding what the context fields are used for.

Reply:
I renamed this type and its fields to hopefully be clearer, and I added a doc comment with an example. The new type name is `TypeContext`.